Process Costing Framework Documentation (#928)
IDAES/idaes-pse: docs/reference_guides/model_libraries/generic/costing/process_costing_sslw.rst

Available IDAES Process Costing Module Methods
----------------------------------------------

When generating a costing block `self.costing` for a unit model `self` (e.g. `m.fs.H101.costing`), the costing block is automatically populated with the following variables:

Table 1. Main Variables added to the unit block ("self.costing").

=========================== ============================== ======== ==========================================================================
Variable                    Symbol                         Units    Notes
=========================== ============================== ======== ==========================================================================
Purchase cost               :math:`purchase\_cost`         dollars  Purchase cost

> Base cost per unit          :math:`base\_cost\_per\_unit`  unitless Base cost per unit

Reviewer: No need to repeat the variable names in the notes column. I'd suggest leaving it blank or even better explaining what purchase cost and base cost are and or the differences between them

    
    m = ConcreteModel()
    m.fs = FlowsheetBlock(dynamic=False)
    
    m.fs.properties = iapws95.Iapws95ParameterBlock()
    
    m.fs.unit = HeatExchanger(
                shell={"property_package": m.fs.properties},
                tube={"property_package": m.fs.properties},
                flow_pattern=HeatExchangerFlowPattern.countercurrent})
    # set inputs
    m.fs.unit.shell_inlet.flow_mol[0].fix(100)     # mol/s
    m.fs.unit.shell_inlet.enth_mol[0].fix(3500)    # j/s
    m.fs.unit.shell_inlet.pressure[0].fix(101325)  # Pa 
    
    m.fs.unit.tube_inlet.flow_mol[0].fix(100)
    m.fs.unit.tube_inlet.enth_mol[0].fix(4000)
    m.fs.unit.tube_inlet.pressure[0].fix(101325.0)
    
    m.fs.unit.area.fix(1000)  # m2
    m.fs.unit.overall_heat_transfer_coefficient.fix(100)  # W/m2K
    
    m.fs.costing = SSLWCosting()
    m.fs.H101.costing = UnitModelCostingBlock(
        flowsheet_costing_block=m.fs.costing,
        costing_method=SSLWCostingData.cost_fired_heater,
        costing_method_arguments={
            "hx_type": HXType.Utube,
            "material_type": HXMaterial.StainlessSteelStainlessSteel,
            "tube_length": HXTubeLength.TwelveFoot,

>             "integer": True,

Reviewer: These look like arguments for `cost_heat_exchanger` not `cost_fired_heater`.

20                     1.00
===================    =====

Construction material correction factor (FM_Mat) can be computed with Eq. 22.44 (Seider et al.)

.. math:: material\_factor = a + (\frac{area}{100})^{b}

Table 5. Materials of construction factors.

================================== ====== ======
Materials of Construction
Shell / Tube                       a      b
================================== ====== ======
carbon steel/carbon steel          0.00   0.00
carbon steel/brass                 1.08   0.05
carbon steel/stainless steel       1.75   0.13
carbon steel/monel                 2.1    0.13
carbon steel/titanium              5.2    0.16
carbon steel/Cr-Mo steel           1.55   0.05
Cr-Mo steel/Cr-Mo steel            1.7    0.07
stainless steel/stainless steel    2.7    0.07
monel/monel                        3.3    0.08
titanium/titanium                  9.6    0.06
================================== ====== ======

Pressure Changer Cost
^^^^^^^^^^^^^^^^^^^^^

The costing of a pressure changer unit model is more complicated, because the pressure changer 
model can be imported into the flowsheet object representing a pump, turbine, compressor, or a 

> simply pressure changer (fan, blower, etc.). The `cost_pressure_changer` method currently supports costing of pumps, turbines, and compressors. The method authomatically interrogates the flowsheet object to determine if the unit is being used as a pump, turbine, or compressor. 

Reviewer: ```suggestion
simple pressure changer (fan, blower, etc.). The `cost_pressure_changer` method currently supports costing of pumps, turbines, and compressors. The method automatically interrogates the flowsheet object to determine if the unit is being used as a pump, turbine, or compressor. 
```

bronze            1.90
stainless steel   2.00
hastelloy C       2.95 
monel             3.30
nickel            3.50
titanium          9.70
================= ======

Electric Motor Pumps
++++++++++++++++++++

A centrifugal pump is usually driven by an electric motor, the `self.costing.motor_purchase_cost` is calculated based on the power consumption and is adjusted by user-defined currency units with the appropriate CE index value.

.. math:: self.costing.motor\_purchase\_cost = FT * self.costing.motor\_base\_cost  (Eq. 22.20)

.. math:: self.costing.motor\_base\_cost = self.costing.motor\_base\_cost\_per\_unit * self.costing.number\_of\_units

.. math:: Q  = self.Q / self.costing.number\_of\_units

.. math:: self.costing.motor\_base\_cost\_per\_unit = e^{(5.8259 + 0.13141\log{PC} + \\ 0.053255(\log{PC})^{2} + 0.028628(\log{PC})^{3} - 0.0035549(\log{PC})^{4})}  (Eq. 22.19)

.. math:: PC = \frac{P_{T}}{\eta_{P}\eta_{M}} = \frac{P_{B}}{\eta_{M}} = \frac{Q H \rho}{33000\eta_{P}\eta_{M}}    (Eq. 22.16)

.. math:: \eta_{P} = -0.316 + 0.24015*\log{Q} - 0.01199(\log{Q})^{2}    (Eq. 22.17)

.. math:: \eta_{M} = 0.80 + 0.0319\log{PB} - 0.00182(\log{PB})^{2}   (Eq. 22.18)

> Efficiencies are valid for PB in the range of 1 to 1500Hp and Q in the range of 50 to 5000 gpm

Reviewer: ```suggestion
Efficiencies are valid for PB in the range of 1 to 1500 Hp and Q in the range of 50 to 5000 gpm
```

Table 11. Driver mover type (for compressors only).

=============== ===============
Mover type      FD (mover_type)
=============== ===============
Electric Mover  1.00
Steam Turbine   1.15
Gas Turbine     1.25
=============== ===============

Table 12. Material of construction factor (for compressors only).

=============== ===========
Material        Mat_factor
=============== ===========
Cast iron       1.00
Stainless steel 1.15
Nickel alloy    1.25
=============== ===========

Fan Cost
++++++++
The fan cost is a function of the actual cubic feet per minute (Q) entering the fan.

Additional arguments are required to estimate the fan cost such as mover_type='fan', fan_head_factor,
fan_type, and material_type ('*' corresponds to the default options):

* fan_type:
    * 'CentrifugalBackward'\*
    * 'CentrifugalStraight'
    * 'VaneAxial'
    * 'TubeAxial'

* fan_head_factor:
    * defaults to 1.45, see table 14

* material_type
    * 'CarbonSteel'
    * 'Fiberglass'
    * 'StainlessSteel'\*
    * 'NickelAlloy'

Additionally, users may pass an argument 'integer' (defaults to True) whether the number of units should be restricted to an integer or not.

> To select the correct fan type users must calculate the total head in inH2O and select the proper fan type from table 13.

Author: The total head is in inches of water ~ I'll write that out explicitly so it's more apparent.

    * `False`

* vessel_diameter:
    * component to use as diameter variable, defaults to unit.diameter

* vessel_length:
    * component to use as length variable, defaults to unit.length

* number_of_units:
    * integer or Pyomo component to use for number of parallel units to be costed, defaults to 1

* number_of_trays:
    * component to use for number of distillation trays in vessel, defaults to `None`

* tray_material:
    * 'CarbonSteel'\*
    * 'StainlessSteel303'
    * 'StainlessSteel316'
    * 'Carpenter20CB3'
    * 'Monel'

* tray_type:
    * 'Sieve'\*
    * 'Valve'
    * 'BubbleCap'

By adding reference parameter, the method can be constructed in any pyomo costing block.
Since the generic models do not include the variables required to cost these type of units, the user must create the blocks and variables.

For example:

.. code:: python

    m.fs.unit = Block()
    m.fs.unit.diameter = Var()
    m.fs.unit.length = Var().
    m.fs.unit.costing = pyo.Block()
    vessel_costing_method = SSLWCostingData.costing_vessel(m.fs.unit.costing, args).

Table 17. Materials of construction factor and material density

================== ====== ================================

> Material Factor    (FM)   methal density (:math:`lb/in^3`)

Reviewer: ```suggestion
Material Factor    (FM)   metal density (:math:`lb/in^3`)
```

================== ====== ================================
carbon_steel       1      0.284
low_alloy_steel    1.2    0.271
stain_steel_304    1.7    0.270
stain_steel_316    2.1    0.276
carpenter_20CB-3   3.2    0.292
nickel_200         5.4    0.3216
monel_400          3.6    0.319
inconel_600        3.9    0.3071
incoloy_825        3.7    0.2903
titanium           7.7    0.1628
================== ====== ================================

Vessel Cost
"""""""""""

> The weight of the unit is calculated based on the methal density, length, Diameter, and shell thickness. `shell_thickness` is a parameter initialized to 1.25, 

Reviewer: ```suggestion
The weight of the unit is calculated based on the metal density, length, diameter, and shell thickness. `shell_thickness` is a parameter initialized to 1.25, 
```

> however, the user must calculate the shell wall minimum thickness computd from the ASME pressure vessel code (tp) add the average vessel thickness, the necessary wall thickness (tE), and select the appropriate shell_thickness.

Reviewer: ```suggestion
however, the user must calculate the shell wall minimum thickness computed from the ASME pressure vessel code (tp) add the average vessel thickness, the necessary wall thickness (tE), and select the appropriate shell_thickness.
```
